activation/challenge_verifier.go

```diff
@@ -0,0 +1,181 @@
+package activation
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+
+	atypes "github.com/spacemeshos/go-spacemesh/activation/types"
+	"github.com/spacemeshos/go-spacemesh/codec"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/signing"
+)
+
+var (
+	ErrSignatureInvalid = errors.New("signature is invalid")
+	ErrChallengeInvalid = errors.New("challenge is invalid")
+	ErrCouldntVerify    = errors.New("couldn't verify challenge")
+)
+
+//go:generate mockgen -package=mocks -destination=./mocks/challenge_verifier.go . AtxProvider
+
+type ChallengeVerificationResult struct {
+	Hash   types.Hash32
+	NodeID types.NodeID
+}
+
+type ChallengeVerifier interface {
+	Verify(ctx context.Context, challenge, signature []byte) (*ChallengeVerificationResult, error)
+}
+
+type AtxProvider interface {
+	GetAtxHeader(id types.ATXID) (*types.ActivationTxHeader, error)
+}
+
+type challengeVerifier struct {
+	atxDB             AtxProvider
+	signatureVerifier signing.VerifyExtractor
+	cfg               atypes.PostConfig
+	goldenATXID       types.ATXID
+}
+
+func NewChallengeVerifier(cdb AtxProvider, signatureVerifier signing.VerifyExtractor, cfg atypes.PostConfig, goldenATX types.ATXID) ChallengeVerifier {
+	return &challengeVerifier{
+		atxDB:             cdb,
+		signatureVerifier: signatureVerifier,
+		cfg:               cfg,
+		goldenATXID:       goldenATX,
+	}
+}
+
+func (v *challengeVerifier) Verify(ctx context.Context, challengeBytes, signature []byte) (*ChallengeVerificationResult, error) {
+	pubkey, err := v.signatureVerifier.Extract(challengeBytes, signature)
+	nodeID := types.BytesToNodeID(pubkey.Bytes())
+	if err != nil {
+		return nil, ErrSignatureInvalid
+	}
+
+	challenge := types.PoetChallenge{}
+	if err := codec.Decode(challengeBytes, &challenge); err != nil {
+		return nil, err
+	}
+
+	if err := v.verifyChallenge(ctx, &challenge, nodeID); err != nil {
+		return nil, err
+	}
+
+	hash, err := challenge.Hash()
+	if err != nil {
+		return nil, err
+	}
+	return &ChallengeVerificationResult{
+		Hash:   *hash,
+		NodeID: nodeID,
+	}, nil
+}
+
+func (v *challengeVerifier) verifyChallenge(ctx context.Context, challenge *types.PoetChallenge, nodeID types.NodeID) error {
+	log.With().Info("verifying challenge", log.Object("challenge", challenge))
+
+	if err := validateNumUnits(&v.cfg, challenge.NumUnits); err != nil {
+		return fmt.Errorf("%w: %v", ErrChallengeInvalid, err)
+	}
+
+	// Verify against positioning ATX
```

Comment:
the verification logic should share code with the atx handler. there should only be some small part that verifies that verifies node's own id that's different

Reply:
Factored out nipost challenge validation.